feat: hyperlink support in engine + hyperlink for path segment

lnu opened the pull request:
### Prerequisites

- [x] I have read and understand the `CONTRIBUTING` guide
- [x] The commit message follows the [conventional commits][cc] guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

### Description

Hyperlink in git and path segment. Supported in windows terminal 1.5(in preview now and available this month for the stable one)

[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

lnu commented:
Work in progress. First draft to see how it works. I have to check how to integrate this cleanly.

JanDeDobbeleer commented:
The goal is that you can click the branch/tag/commit and link to the repo?

lnu commented:
> The goal is that you can click the branch/tag/commit and link to the repo?

Yes it's a first draft to see how it looks like.  
Right now the ansi sequence is embedded in the segment for testing purpose. I was thinking of using the markdown syntax for url to decouple the ansi rendering from the segment. I have a more or less working regex but it's not easy to cover all scenarios.

tradiff commented on src/segment_git.go:
```diff
@@ -131,12 +135,23 @@ func (g *git) string() string {
+	}
+	if g.repo.upstream != "" && enableHyperlink {
+		url := strings.Replace(g.repo.url, ".git", "", -1)
+		fmt.Fprintf(buffer, "]8;;%s\\%s]8;;\\", fmt.Sprintf("%s/tree/%s", url, g.repo.HEAD), g.repo.PRETTYHEAD)
```
The `/tree/` url seems github specific. At my work, we use Azure Devops and the url is different there. Your code results in a 404 page for my Azure Devops projects.

Example:
git remote get-url origin:
`https://AcmeSoftware@dev.azure.com/AcmeSoftware/OhMyPosh/_git/oh-my-posh3`
branch:
`feat/hyperlink`
url:
`https://AcmeSoftware@dev.azure.com/AcmeSoftware/OhMyPosh/_git/oh-my-posh3?version=GBfeat%2Fmy-feature`

 The url construction code would be something like this (untested):
```go
fmt.Sprintf("%s?version=GB%s", url, g.repo.HEAD)
```

Maybe we could do a check to see if the url contains `github.com` or `dev.azure.com` and apply the proper url building? And for other domains we don't know about, just omit the hyperlink feature. Or make the hyperlink point to the git repo url directly (no branch)

lnu replied:
I was expecting some issues like that. Yeah of course let's keep a list of main providers and point to the repo when unknown.
Thanks for the feedback!

lnu replied:
I did a first draft of multi providers support, if you have the opportunity to test it.

tradiff commented:
>  Supported in windows terminal 1.5(in preview now and available this month for the stable one)

BTW I'm using windows terminal 1.4.3243.0 and the clickable link works for me ¯\\\_(ツ)_/¯

tradiff commented on src/segment_git.go:
```diff
@@ -131,12 +146,38 @@ func (g *git) string() string {
+			}
+		}
+		if urlTemplate != "" {
+			// if provider found, open the branch
```
```suggestion
			// if provider found, open the branch

			// normalize azure devops url to change from
			// https://{account}@dev.azure.com/{account}/{project}/_git/{repo}
			// to
			// https://dev.azure.com/{account}/{project}/_git/{repo}
			repoURL = replaceAllString("//[^@]*@", repoURL, "//")						
```

This suggestion makes the code work for me, for azure devops repositories.  I suspect you'll also have problems with people who have cloned using ssh instead of https. I found a couple other repos which are doing something similar to what you're doing, which may be useful to review.

[dotnet/docfx](https://github.com/dotnet/docfx/blob/dev/src/docfx.website.themes/common/common.js) is probably the easiest to follow
[eamodio/vscode-gitlens](https://github.com/eamodio/vscode-gitlens/tree/main/src/git/remotes) has this functionality too

Hope this helps!

lnu replied:
Thanks. I thought the host property was stripping the account and returning the hostname only. I’ll check this.

lnu commented:
I still have an issue for path not rendered correctly when mapped_locations is used and a unicode character is set.
Example:
```
{
          "type": "path",
          "style": "powerline",
          "powerline_symbol": "\uE0B0",
          "foreground": "#ffffff",
          "background": "#ff479c",
          "properties": {
            "prefix": " \uE5FF ",
            "style": "full",
            "mapped_locations": [
              ["D:\\dev", "\uf962 "]
            ],
            "mapped_locations_enabled": true,
            "enable_hyperlink": true
          }
}
```
rendering:
![image](https://user-images.githubusercontent.com/1829553/103647496-b27bad00-4f5b-11eb-855c-632561ec7978.png)

With something else:
![image](https://user-images.githubusercontent.com/1829553/103647624-e48d0f00-4f5b-11eb-9298-9c0362209cdb.png)


I think it can be related to that windows terminal issue: https://github.com/microsoft/terminal/issues/3546

lnu commented:
and support for opening uri is coming:https://github.com/microsoft/terminal/pull/7526
we're going too fast 🤣 

to track hyperlink progress: https://github.com/microsoft/terminal/issues/5001

lnu commented:
And  to test hyperlink in the terminal:
`printf '\e]8;;https://github.com/lnu/oh-my-posh3.git\e\\main\e]8;;\e\\'`

JanDeDobbeleer commented:
@lnu I was looking for that unicode bug the other day. Sometimes I think this whole repo is "too soon" for MS as we're constantly battling with incomplete features whereas users expect all of this to "just work". Or, we're right on time and will provide appropriate focus 😅

lnu commented:
tested in gnome terminal and everything works(uri + http), and no glitch:
![image](https://user-images.githubusercontent.com/1829553/103780105-d6f78800-5034-11eb-90c7-c51cd07e53d9.png)

JanDeDobbeleer commented on src/segment_git.go:
```diff
@@ -3,19 +3,32 @@ package main
+		"github.com":       "%s/tree/%s",
+		"azure.com":        "%s?version=GB%s",
+		"visualstudio.com": "%s?version=GB%s",
+		"gitlab.com":       "%s/-/tree/%s",
```
we should also add bitbucket, and that said, add support for these remotes too while we're at it. Maybe that got me thinking, this can be part of that configuration. The remote url match is hardcoded but can be rewritten and easily extended this way. Maybe that's out of scope for this change, but still a good idea.

lnu replied:
resolved by moving git segment to another pull request

JanDeDobbeleer commented on src/segment_git.go:
```diff
@@ -131,12 +146,44 @@ func (g *git) string() string {
+	if g.repo.upstream != "" && enableHyperlink {
+		// parse url to and strips extra info
+		baseURL, err := url.Parse(g.repo.url)
+		if err == nil {
```
Do an early return here rather than embedding it in a whole block.

JanDeDobbeleer commented on src/segment_git.go:
```diff
@@ -131,12 +146,44 @@ func (g *git) string() string {
+	// default name
+	repoHEAD := g.repo.PRETTYHEAD
+	if g.repo.upstream != "" && enableHyperlink {
+		// parse url to and strips extra info
```
this whole logic should be a separate function

lnu replied:
I think we should handle hyperlink at a global level(https://github.com/JanDeDobbeleer/oh-my-posh3/pull/307#issuecomment-753464224).  
Outputting something like a template and detect it in the ansi renderer.

JanDeDobbeleer replied:
Yes!

lnu replied:
Now that i’m back to work I have less time but I’ll try do finish it by the end of the week.
Url template
Global hyperlink
Extra: support ssh cloned repo

JanDeDobbeleer commented on src/segment_git.go:
```diff
@@ -131,12 +146,44 @@ func (g *git) string() string {
+			if urlTemplate != "" {
+				// if provider found, open the branch
+				repoHEAD = fmt.Sprintf("]8;;%s\\%s]8;;\\", fmt.Sprintf(urlTemplate, repoURL, g.repo.HEAD), g.repo.PRETTYHEAD)
+			} else {
```
once this logic is in a separate function, the else can be removed when you return a string and write this in the function above

JanDeDobbeleer commented on src/segment_git.go:
```diff
@@ -131,12 +146,44 @@ func (g *git) string() string {
+			repoURL := fmt.Sprintf("%s://%s%s", baseURL.Scheme, baseURL.Host, strings.ReplaceAll(baseURL.Path, ".git", ""))
+			if urlTemplate != "" {
+				// if provider found, open the branch
+				repoHEAD = fmt.Sprintf("]8;;%s\\%s]8;;\\", fmt.Sprintf(urlTemplate, repoURL, g.repo.HEAD), g.repo.PRETTYHEAD)
```
the formatting needs to become part of `ansi_formats.go`. Wondering if we want to expose this just like we do for color rendering overrides?

lnu replied:
yes of course => https://github.com/JanDeDobbeleer/oh-my-posh3/pull/307#issuecomment-753464224

lnu commented:
I did some refactoring regarding the hyperlink support. I think we should split the functionality in two:
- this pull request with:
  - hyperlink support by the engine using url markdown syntax
  - Hyperlink support for path segment
  - Initial hyperlink support for git segment, limited to http cloned repo
- a second pull with 
  - A refactoring of the git url parsing (like it's done in git lens for ex.)
  - Support for ssl cloned repo
  - Support for custom config

Otherwise, it will take too long to have a first working version. What do you think @JanDeDobbeleer ?

JanDeDobbeleer commented:
@lnu I would then maybe, for the sake of simplicity, only limit to the engine and path segment for 1. Leave git on a separate track.

lnu committed:
feat: add hyperlink rendering

An hyperlink can be added using markdown syntax and will be detected by the engine.
Initial implementation for path segments.

JanDeDobbeleer commented on src/ansi_formats.go:
```diff
@@ -69,6 +70,7 @@ func (a *ansiFormats) init(shell string) {
 		a.escapeLeft = ""
 		a.escapeRight = ""
 	}
+	a.hyperlink = "\x1b]8;;%s\x1b\\%s\x1b]8;;\x1b\\"
```
this needs to be escaped differently for zsh and bash, otherwise the line of the prompt won't be correct anymore and rendering is ruined. You can validate this by doing a tab completion using `cd` for example and notice the prompt just a bit off after pressing `TAB` the first time.

JanDeDobbeleer commented on src/engine.go:
```diff
@@ -82,6 +83,16 @@ func (e *engine) renderText(text string) {
 	}
+
+	// hyperlink matching
+	results := findNamedRegexMatch("(?P<all>(?:\\[(?P<name>.+)\\])(?:\\((?P<url>.*)\\)))", text)
```
I would make this a function in `ansi_formats.go` so it can also be tested properly.

JanDeDobbeleer commented on src/engine.go:
```diff
@@ -166,7 +176,7 @@ func (e *engine) render() {
 }
 
-// debug will lool through your config file and output the timings for each segments
+// debug will loop through your config file and output the timings for each segments
```
that made me laugh actually :laughing

JanDeDobbeleer commented on src/engine.go:
```diff
@@ -217,6 +227,7 @@ func (e *engine) debug() {
 		segmentName := fmt.Sprintf("%s(%t)", segment.name, segment.enabled)
 		e.renderer.print(fmt.Sprintf("%-*s - %3d ms - %s\n", largestSegmentNameLength, segmentName, duration, segment.stringValue))
 	}
+
```
unneeded change unless our listing expects this

JanDeDobbeleer commented on src/settings.go:
```diff
@@ -14,6 +14,7 @@ type Settings struct {
 	ConsoleTitle         bool              `json:"console_title"`
 	ConsoleTitleStyle    ConsoleTitleStyle `json:"console_title_style"`
 	ConsoleTitleTemplate string            `json:"console_title_template"`
+	EnableHyperlink      bool              `json:"enable_hyperlink"`
```
Is there a side-effect of having this enabled but not supported in your terminal?

lnu replied:
yes I think your prompt won't be happy if the terminal you use don't support hyperlink. I read various threads and it seems there's no way to know if the terminal supports it or not. So, by default it's disabled.

lnu replied:
If you check [here](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda), a majority of terminal already support it.

lnu commented:
Still have to change the call in engine.go to generateHyperlink. Calling through color is a bit strange.

JanDeDobbeleer commented:
> Calling through color is a bit strange

I agree, but `ansi_formats.go` makes sense.

JanDeDobbeleer reviewed (changes requested):
What I'm missing is an addition to the docs about this like colors. Because I can create a text segment/prefix/postfix, that contains a link which will be rendered when hyperlinks are enabled.

lnu commented:
> What I'm missing is an addition to the docs about this like colors. Because I can create a text segment/prefix/postfix, that contains a link which will be rendered when hyperlinks are enabled.

Yes but be careful that right now the engine replaces only hyperlink in segment text, not prefix or postfix.

JanDeDobbeleer commented:
> Yes but be careful that right now the engine replaces only hyperlink in segment text, not prefix or postfix.

True. Which is fine I guess, this needs to be more contextual anyways. It's not because we can that we should.

JanDeDobbeleer commented on docs/docs/segment-path.md:
```diff
@@ -37,6 +37,7 @@ Display the current path.
 is set to `true`)
 - mapped_locations_enabled: `boolean` - replace known locations in the path with the replacements before applying the
 style. defaults to `true`
+- enable_hyperlink: `boolean` - displays an hyperlink for the path - defaults to `false`
```
Can't believe I missed this, do we want to keep it at this level or higher up at `Settings`?

lnu replied:
Now, the engine detects url from any segments. Hyperlink can be enabled/disabled for each segment and it takes care of rendering the text or the url markdown.  
On the other side, it could be enabled at global level and implying that all segments supporting url will render them.
Both have their pros and cons.  
It’s a matter of taste :).

JanDeDobbeleer replied:
I believe both can be an option, but let's keep it like this for now. Amazing work.

imba-tjd commented on docs/docs/configuration.md:
```diff
@@ -291,6 +291,19 @@ Oh my Posh mainly supports three different color types being
+
+#### Supported terminals
+
+- [Terminal list](thttps://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda)
```
`t`

JanDeDobbeleer replied:
fixed, thx.
